File: hs-src/GitRestack/Cleanup.hs
{-# LANGUAGE OverloadedStrings #-}

module GitRestack.Cleanup
  ( CleanupOptions(..)
  , runCleanup
  ) where

import Control.Monad (forM_, when)
import qualified Data.ByteString as BS
import qualified Data.ByteString.Char8 as BSC
import Data.Maybe (fromMaybe)
import qualified Data.Text as Text
import GitRestack.Git
import GitRestack.Types
import GitRestack.Utils (isFixBranch, makeFixBranchName)
import GitRestack.Yaml
import System.Directory
import System.Exit (ExitCode(..), exitWith)
import System.FilePath (takeFileName)


data CleanupOptions = CleanupOptions
  { coPlanFile :: Maybe FilePath
  , coWorktreePath :: Maybe FilePath
  , coForce :: Bool
  , coKeepPlan :: Bool
  }

runCleanup :: CleanupOptions -> IO ()
runCleanup opts = do
  cwd <- getCwd
  let repoName = takeFileName cwd
  let defaultWorktree = "../" <> repoName <> "-restack"
  let worktreePath = fromMaybe defaultWorktree (coWorktreePath opts)

  let planPath = fromMaybe ".git/git-restack/plan.yml" (coPlanFile opts)

  planInfo <- loadPlanInfo planPath
  let planWorktreePath = planInfoWorktree planInfo
  let fixBranches = planInfoFixBranches planInfo
  let backupBranches = planInfoBackupBranches planInfo
  let planBranches = planInfoPlanBranches planInfo

  hasWorktree <- doesPathExist worktreePath
  hasPlanWorktree <- maybe (pure False) doesPathExist planWorktreePath

  let hasPlanFile = planInfoHasPlan planInfo
  let hasStateFile = planInfoHasState planInfo

  when (not hasWorktree && not hasPlanWorktree && null fixBranches && null backupBranches && null planBranches && not hasPlanFile && not hasStateFile) $ do
    putStrLn "No git-restack artifacts found."
    exitWith ExitSuccess

  putStrLn "\nCleanup will remove:\n"
  when hasWorktree $ putStrLn ("  - Worktree: " <> worktreePath)
  when hasPlanWorktree $ putStrLn ("  - Plan worktree: " <> fromMaybe "" planWorktreePath)

  when (not (null fixBranches)) $ do
    putStrLn ("  - " <> show (length fixBranches) <> " git-restack/fix branches:")
    forM_ fixBranches $ \branch -> putStrLn ("    - " <> branch)

  when (not (null backupBranches)) $ do
    putStrLn ("  - " <> show (length backupBranches) <> " backup branches:")
    forM_ backupBranches $ \branch -> putStrLn ("    - " <> branch)

  when (not (null planBranches)) $ do
    putStrLn ("  - " <> show (length planBranches) <> " plan branches:")
    forM_ planBranches $ \branch -> putStrLn ("    - " <> branch)

  when (not (coKeepPlan opts)) $ do
    when hasPlanFile $ putStrLn ("  - Plan file: " <> planPath)
    when hasStateFile $ putStrLn "  - State file: .git/git-restack/state.json"

  putStrLn ""

  unlessForce (coForce opts) $ do
    putStrLn "Run with --force to confirm cleanup."
    exitWith ExitSuccess

  putStrLn "\nCleaning up...\n"

  when hasWorktree $ removeWorktree worktreePath
  when hasPlanWorktree $ do
    let planPathToRemove = fromMaybe worktreePath planWorktreePath
    when (planPathToRemove /= worktreePath) $ removeWorktree planPathToRemove

  _ <- runGitWithStatus ["worktree", "prune"]

  deleteBranches fixBranches "Deleted"
  deleteBranches backupBranches "Deleted backup"
  deleteBranches planBranches "Deleted plan"

  when (not (coKeepPlan opts)) $ do
    removeFileIfExists ".git/git-restack/state.json"
    removeFileIfExists planPath
    removeDirIfExists ".git/git-restack"
    putStrLn "  [OK] Removed state files"

  putStrLn "\nDone! Cleanup complete."

unlessForce :: Bool -> IO () -> IO ()
unlessForce forceFlag action = when (not forceFlag) action

removeWorktree :: FilePath -> IO ()
removeWorktree path = do
  result <- runGitWithStatus ["worktree", "remove", path, "--force"]
  if grExitCode result == 0
    then putStrLn ("  [OK] Removed worktree: " <> path)
    else do
      removePathForcibly path
      putStrLn ("  [OK] Removed worktree (forced): " <> path)

removeFileIfExists :: FilePath -> IO ()
removeFileIfExists path = do
  exists <- doesFileExist path
  when exists (removeFile path)

removeDirIfExists :: FilePath -> IO ()
removeDirIfExists path = do
  exists <- doesDirectoryExist path
  when exists (removeDirectoryRecursive path)


data PlanInfo = PlanInfo
  { planInfoFixBranches :: [String]
  , planInfoBackupBranches :: [String]
  , planInfoPlanBranches :: [String]
  , planInfoWorktree :: Maybe FilePath
  , planInfoHasPlan :: Bool
  , planInfoHasState :: Bool
  }

loadPlanInfo :: FilePath -> IO PlanInfo
loadPlanInfo planPath = do
  hasPlan <- doesFileExist planPath
  hasState <- doesFileExist ".git/git-restack/state.json"
  if hasPlan
    then do
      content <- BS.readFile planPath
      case parsePlan content of
        Right plan -> do
          let fixBranches = map (Text.unpack . makeFixBranchName . sbName) (stBranches (planStack plan))
          let (planWorktree, planBranches, backupBranches) =
                case planSimulation plan of
                  Nothing -> (Nothing, [], [])
                  Just sim ->
                    ( simWorktreePath sim
                    , maybe [] (\name -> [Text.unpack name]) (simPlanBranch sim)
                    , map (Text.unpack . backupName) (simBackupBranches sim)
                    )
          pure PlanInfo
            { planInfoFixBranches = fixBranches
            , planInfoBackupBranches = backupBranches
            , planInfoPlanBranches = planBranches
            , planInfoWorktree = planWorktree
            , planInfoHasPlan = True
            , planInfoHasState = hasState
            }
        Left _ -> fallbackPlanInfo hasPlan hasState
    else fallbackPlanInfo hasPlan hasState
  where
    fallbackPlanInfo hasPlan hasState = do
      fixBranches <- listBranches "git-restack/fix/*" (\name -> isFixBranch (Text.pack name))
      backupBranches <- listBranches "*-restack-backup-*" (\name -> "restack-backup" `Text.isInfixOf` Text.pack name)
      planBranches <- listBranches "git-restack-plan-*" (\name -> "git-restack-plan-" `Text.isPrefixOf` Text.pack name)
      pure PlanInfo
        { planInfoFixBranches = fixBranches
        , planInfoBackupBranches = backupBranches
        , planInfoPlanBranches = planBranches
        , planInfoWorktree = Nothing
        , planInfoHasPlan = hasPlan
        , planInfoHasState = hasState
        }

listBranches :: String -> (String -> Bool) -> IO [String]
listBranches pattern predicate = do
  result <- runGitWithStatus ["branch", "--list", pattern]
  if grExitCode result /= 0
    then pure []
    else pure (filter predicate (map parseBranchLine (lines (BSC.unpack (grStdout result)))))
  where
    parseBranchLine line =
      let trimmed = dropWhile (== ' ') (dropWhile (== '*') line)
      in dropWhile (== ' ') trimmed


deleteBranches :: [String] -> String -> IO ()
deleteBranches branches label =
  forM_ branches $ \branch -> do
    result <- runGitWithStatus ["branch", "-D", branch]
    if grExitCode result == 0
      then putStrLn ("  [OK] " <> label <> ": " <> branch)
      else putStrLn ("  [WARN] Could not delete: " <> branch)

File: src/cli/cleanup.zig
const std = @import("std");
const process = @import("../utils/process.zig");
const strings = @import("../utils/strings.zig");
const restack = @import("../utils/restack.zig");
const parser = @import("../yaml/parser.zig");

const STATE_DIR = ".git/git-restack";
const STATE_FILE = ".git/git-restack/state.json";
const PLAN_FILE = ".git/git-restack/plan.yml";

pub fn run(allocator: std.mem.Allocator, args: []const []const u8) !void {
    var plan_file: ?[]const u8 = null;
    var worktree_path: ?[]const u8 = null;
    var force = false;
    var keep_plan = false;

    // Parse options
    var i: usize = 0;
    while (i < args.len) : (i += 1) {
        const arg = args[i];
        if (std.mem.eql(u8, arg, "--worktree-path")) {
            if (i + 1 < args.len) {
                i += 1;
                worktree_path = args[i];
            }
        } else if (std.mem.eql(u8, arg, "-f") or std.mem.eql(u8, arg, "--force")) {
            force = true;
        } else if (std.mem.eql(u8, arg, "--keep-plan")) {
            keep_plan = true;
        } else if (std.mem.eql(u8, arg, "-h") or std.mem.eql(u8, arg, "--help")) {
            printHelp();
            return;
        } else if (!std.mem.startsWith(u8, arg, "-")) {
            plan_file = arg;
        }
    }

    // Determine worktree path
    const cwd = try process.getCwd(allocator);
    defer allocator.free(cwd);
    const repo_name = std.fs.path.basename(cwd);
    const default_worktree = try std.fmt.allocPrint(allocator, "../{s}-restack", .{repo_name});
    defer allocator.free(default_worktree);
    const wt_path = worktree_path orelse default_worktree;

    // Check if worktree exists
    const has_worktree = blk: {
        std.fs.cwd().access(wt_path, .{}) catch break :blk false;
        break :blk true;
    };

    // Try to load plan to find fix/backup/plan branches
    var fix_branches: std.ArrayListUnmanaged([]const u8) = .{};
    var backup_branches: std.ArrayListUnmanaged([]const u8) = .{};
    var plan_branches: std.ArrayListUnmanaged([]const u8) = .{};
    var plan_worktree_path: ?[]const u8 = null;
    defer {
        for (fix_branches.items) |b| allocator.free(b);
        fix_branches.deinit(allocator);
        for (backup_branches.items) |b| allocator.free(b);
        backup_branches.deinit(allocator);
        for (plan_branches.items) |b| allocator.free(b);
        plan_branches.deinit(allocator);
        if (plan_worktree_path) |path| allocator.free(path);
    }

    // Try default plan location, then explicit plan file
    const plan_path = plan_file orelse PLAN_FILE;
    
    if (std.fs.cwd().readFileAlloc(allocator, plan_path, 10 * 1024 * 1024)) |plan_content| {
        defer allocator.free(plan_content);

        if (parser.parsePlan(allocator, plan_content)) |plan_value| {
            var plan = plan_value;
            defer plan.deinit(allocator);

            for (plan.stack.branches) |branch| {
                const fix_name = restack.makeFixBranchName(allocator, branch.name) catch continue;
                fix_branches.append(allocator, fix_name) catch continue;
            }

            if (plan.simulation) |sim| {
                if (sim.worktree_path) |path| {
                    plan_worktree_path = strings.copy(allocator, path) catch null;
                }
                if (sim.plan_branch) |branch_name| {
                    if (strings.copy(allocator, branch_name)) |copied| {
                        plan_branches.append(allocator, copied) catch allocator.free(copied);
                    } else |_| {}
                }
                for (sim.backup_branches) |backup| {
                    if (strings.copy(allocator, backup.backup)) |copied| {
                        backup_branches.append(allocator, copied) catch allocator.free(copied);
                    } else |_| {}
                }
            }
        } else |_| {}
    } else |_| {
        // No plan file, find fix branches by pattern
        const branches_result = process.runGitWithStatus(allocator, &.{
            "branch",
            "--list",
            "git-restack/fix/*",
        }) catch null;

        if (branches_result) |result| {
            defer allocator.free(result.stdout);
            defer allocator.free(result.stderr);

            if (result.exit_code == 0) {
                var lines = std.mem.splitScalar(u8, result.stdout, '\n');
                while (lines.next()) |line| {
                    const trimmed = strings.trim(line);
                    const branch_name = if (trimmed.len > 2 and (trimmed[0] == '*' or trimmed[0] == ' '))
                        strings.trim(trimmed[2..])
                    else
                        trimmed;
                    
                    if (branch_name.len > 0 and restack.isFixBranch(branch_name)) {
                        fix_branches.append(allocator, strings.copy(allocator, branch_name) catch continue) catch continue;
                    }
                }
            }
        }

        const backup_result = process.runGitWithStatus(allocator, &.{
            "branch",
            "--list",
            "*-restack-backup-*",
        }) catch null;

        if (backup_result) |result| {
            defer allocator.free(result.stdout);
            defer allocator.free(result.stderr);

            if (result.exit_code == 0) {
                var lines = std.mem.splitScalar(u8, result.stdout, '\n');
                while (lines.next()) |line| {
                    const trimmed = strings.trim(line);
                    const branch_name = if (trimmed.len > 2 and (trimmed[0] == '*' or trimmed[0] == ' '))
                        strings.trim(trimmed[2..])
                    else
                        trimmed;

                    if (branch_name.len > 0 and std.mem.indexOf(u8, branch_name, "restack-backup") != null) {
                        backup_branches.append(allocator, strings.copy(allocator, branch_name) catch continue) catch continue;
                    }
                }
            }
        }

        const plan_branch_result = process.runGitWithStatus(allocator, &.{
            "branch",
            "--list",
            "git-restack-plan-*",
        }) catch null;

        if (plan_branch_result) |result| {
            defer allocator.free(result.stdout);
            defer allocator.free(result.stderr);

            if (result.exit_code == 0) {
                var lines = std.mem.splitScalar(u8, result.stdout, '\n');
                while (lines.next()) |line| {
                    const trimmed = strings.trim(line);
                    const branch_name = if (trimmed.len > 2 and (trimmed[0] == '*' or trimmed[0] == ' '))
                        strings.trim(trimmed[2..])
                    else
                        trimmed;

                    if (branch_name.len > 0 and std.mem.startsWith(u8, branch_name, "git-restack-plan-")) {
                        plan_branches.append(allocator, strings.copy(allocator, branch_name) catch continue) catch continue;
                    }
                }
            }
        }
    }

    const has_plan_worktree = if (plan_worktree_path) |path| blk: {
        std.fs.cwd().access(path, .{}) catch break :blk false;
        break :blk true;
    } else false;

    if (!has_worktree and !has_plan_worktree and fix_branches.items.len == 0 and backup_branches.items.len == 0 and plan_branches.items.len == 0) {
        std.debug.print("No git-restack artifacts found.\n", .{});
        return;
    }

    // Show what will be deleted
    std.debug.print("\nCleanup will remove:\n\n", .{});
    if (has_worktree) {
        std.debug.print("  • Worktree: {s}\n", .{wt_path});
    }
    if (has_plan_worktree) {
        std.debug.print("  • Plan worktree: {s}\n", .{plan_worktree_path.?});
    }
    
    if (fix_branches.items.len > 0) {
        std.debug.print("  • {d} git-restack/fix branches:\n", .{fix_branches.items.len});
        for (fix_branches.items) |branch| {
            std.debug.print("    └── {s}\n", .{branch});
        }
    }

    if (backup_branches.items.len > 0) {
        std.debug.print("  • {d} backup branches:\n", .{backup_branches.items.len});
        for (backup_branches.items) |branch| {
            std.debug.print("    └── {s}\n", .{branch});
        }
    }

    if (plan_branches.items.len > 0) {
        std.debug.print("  • {d} plan branches:\n", .{plan_branches.items.len});
        for (plan_branches.items) |branch| {
            std.debug.print("    └── {s}\n", .{branch});
        }
    }

    if (!keep_plan) {
        const has_plan = blk: {
            std.fs.cwd().access(PLAN_FILE, .{}) catch break :blk false;
            break :blk true;
        };
        if (has_plan) {
            std.debug.print("  • Plan file: {s}\n", .{PLAN_FILE});
        }
        
        const has_state = blk: {
            std.fs.cwd().access(STATE_FILE, .{}) catch break :blk false;
            break :blk true;
        };
        if (has_state) {
            std.debug.print("  • State file: {s}\n", .{STATE_FILE});
        }
    }

    std.debug.print("\n", .{});

    // Require --force for non-interactive use
    if (!force) {
        std.debug.print("Run with --force to confirm cleanup.\n", .{});
        return;
    }

    std.debug.print("\nCleaning up...\n\n", .{});

    if (has_worktree) {
        const wt_result = process.runGitWithStatus(allocator, &.{
            "worktree",
            "remove",
            wt_path,
            "--force",
        }) catch null;

        if (wt_result) |r| {
            defer allocator.free(r.stdout);
            defer allocator.free(r.stderr);

            if (r.exit_code == 0) {
                std.debug.print("  \x1b[32m✓\x1b[0m Removed worktree\n", .{});
            } else {
                std.fs.cwd().deleteTree(wt_path) catch {};
                std.debug.print("  \x1b[32m✓\x1b[0m Removed worktree (forced)\n", .{});
            }
        }
    }

    if (has_plan_worktree and (plan_worktree_path == null or !std.mem.eql(u8, plan_worktree_path.?, wt_path))) {
        const plan_wt_path = plan_worktree_path.?;
        const plan_result = process.runGitWithStatus(allocator, &.{
            "worktree",
            "remove",
            plan_wt_path,
            "--force",
        }) catch null;

        if (plan_result) |r| {
            defer allocator.free(r.stdout);
            defer allocator.free(r.stderr);

            if (r.exit_code == 0) {
                std.debug.print("  \x1b[32m✓\x1b[0m Removed plan worktree\n", .{});
            } else {
                std.fs.cwd().deleteTree(plan_wt_path) catch {};
                std.debug.print("  \x1b[32m✓\x1b[0m Removed plan worktree (forced)\n", .{});
            }
        }
    }

    _ = process.runGit(allocator, &.{ "worktree", "prune" }) catch {};

    // Delete fix branches
    for (fix_branches.items) |branch| {
        const result = process.runGitWithStatus(allocator, &.{
            "branch",
            "-D",
            branch,
        }) catch continue;
        defer allocator.free(result.stdout);
        defer allocator.free(result.stderr);

        if (result.exit_code == 0) {
            std.debug.print("  \x1b[32m✓\x1b[0m Deleted: {s}\n", .{branch});
        } else {
            std.debug.print("  \x1b[33m⚠\x1b[0m Could not delete: {s}\n", .{branch});
        }
    }

    for (backup_branches.items) |branch| {
        const result = process.runGitWithStatus(allocator, &.{
            "branch",
            "-D",
            branch,
        }) catch continue;
        defer allocator.free(result.stdout);
        defer allocator.free(result.stderr);

        if (result.exit_code == 0) {
            std.debug.print("  \x1b[32m✓\x1b[0m Deleted backup: {s}\n", .{branch});
        } else {
            std.debug.print("  \x1b[33m⚠\x1b[0m Could not delete backup: {s}\n", .{branch});
        }
    }

    for (plan_branches.items) |branch| {
        const result = process.runGitWithStatus(allocator, &.{
            "branch",
            "-D",
            branch,
        }) catch continue;
        defer allocator.free(result.stdout);
        defer allocator.free(result.stderr);

        if (result.exit_code == 0) {
            std.debug.print("  \x1b[32m✓\x1b[0m Deleted plan branch: {s}\n", .{branch});
        } else {
            std.debug.print("  \x1b[33m⚠\x1b[0m Could not delete plan branch: {s}\n", .{branch});
        }
    }

    // Delete state files
    if (!keep_plan) {
        std.fs.cwd().deleteFile(STATE_FILE) catch {};
        std.fs.cwd().deleteFile(PLAN_FILE) catch {};
        std.fs.cwd().deleteDir(STATE_DIR) catch {};
        std.debug.print("  \x1b[32m✓\x1b[0m Removed state files\n", .{});
    }

    std.debug.print("\n\x1b[32mDone!\x1b[0m Cleanup complete.\n", .{});
}

fn printHelp() void {
    std.debug.print(
        \\Usage: git-restack cleanup [plan.yml] [OPTIONS]
        \\
        \\Removes git-restack worktrees and temporary branches after you're done reviewing.
        \\Run this after 'git-restack apply' when you're satisfied with the results.
        \\
        \\This removes:
        \\  - The worktree (../<repo>-restack)
        \\  - The plan worktree (../<repo>-restack-plan)
        \\  - All git-restack/fix/* branches created by 'exec'
        \\  - Any git-restack plan/backup branches
        \\  - The plan and state files (unless --keep-plan)
        \\
        \\Options:
        \\  --worktree-path <path>   Path to worktree (default: ../<repo>-restack)
        \\  --keep-plan              Don't delete the plan file
        \\  -f, --force              Skip confirmation prompt
        \\  -h, --help               Show this help message
        \\
        \\Examples:
        \\  git-restack cleanup                 # Clean up default locations
        \\  git-restack cleanup --force         # Skip confirmation
        \\  git-restack cleanup --keep-plan     # Keep plan for reference
        \\
    , .{});
}
